fix(core): NgZone coaleascing options should trigger onStable correctly

JiaLiPassion opened the pull request:
fix https://github.com/angular/components/issues/21674

When setting `ngZoneRunCoalescing` to true, `onStable` is not emitted correctly.
The reason is before this commit, the code looks like is,

```
// Application code call `ngZone.run()`
ngZone.run(() => {}); // step 1

// Inside NgZone, in the OnInvoke hook, NgZone try to delay the checkStable()

function delayChangeDetectionForEvents(zone: NgZonePrivate) {
  if (zone.lastRequestAnimationFrameId !== -1) { // step 9
    return;
  }
  zone.lastRequestAnimationFrameId = zone.nativeRequestAnimationFrame.call(global, () => { // step 2
    if (!zone.fakeTopEventTask) {
      zone.fakeTopEventTask = Zone.root.scheduleEventTask('fakeTopEventTask', () => {
        zone.lastRequestAnimationFrameId = -1; // step 3
        updateMicroTaskStatus(zone); // step 4
        checkStable(zone); // step 6
      }, undefined, () => {}, () => {});
    }
    zone.fakeTopEventTask.invoke();
  });
  updateMicroTaskStatus(zone);
}

function updateMicroTaskStatus(zone: NgZonePrivate, ignoreCheckRAFId = false) {
  if (zone._hasPendingMicrotasks ||
      ((zone.shouldCoalesceEventChangeDetection || zone.shouldCoalesceRunChangeDetection) &&
       zone.lastRequestAnimationFrameId !== -1)) { // step 5
    zone.hasPendingMicrotasks = true;
  } else {
    zone.hasPendingMicrotasks = false;
  }
}

function checkStable(zone: NgZonePrivate) {
  if (zone._nesting == 0 && !zone.hasPendingMicrotasks && !zone.isStable) { // step 7
    try {
      zone._nesting++;
      zone.onMicrotaskEmpty.emit(null);
    ...
}

// application ref subscribe onMicroTaskEmpty
ngZone.onMicroTaskEmpty.subscribe(() => {
  ngZone.run(() => { // step 8
    tick();
  });
});

```

And the process is:
1. step 1: application call ngZone.run()
2. step 2: NgZone delay the checkStable() call in a requestAnimationFrame, and also set
zone.lastRequestAnimationFrameId
3. step 3: Inside the requestAnimationFrame callback, reset zone.lastRequestAnimationFrameId first
4. step 4: update microTask status
5, step 5: if zone.lastRequestAnimationFrameId is -1, that means no microTask pending.
6. step 6: checkStable and trigger onMicrotaskEmpty emitter.
7. step 7: ApplicationRef subscribed onMicrotaskEmpty, so it will call another `ngZone.run()` to process
tick()
8. step 8: And this new `ngZone.run()` will try to check `zone.lastRequestAnimationFrameId` in `step 9`
when trying to delay the checkStable(), and since the zone.lastRequestAnimationFrameId is already reset
to -1 in step 3, so this ngZone.run() will run into step 2 again.
9. And become a infinite loop..., so onStable is never emit

In this commit, the `zone.lastRequestAnimationFrameId` reset is moved after `checkStable()` call.

mhevery commented on packages/core/src/zone/ng_zone.ts:
```diff
@@ -243,6 +243,7 @@ interface NgZonePrivate extends NgZone {
   hasPendingMacrotasks: boolean;
   hasPendingMicrotasks: boolean;
   lastRequestAnimationFrameId: number;
+  isRunningDelayChangeDetection: boolean;
```
I see this declared in the `NgZonePrivate` interface but I don't see it initialized in the `NgZone`. Can you add initialization?

JiaLiPassion replied:
Sure, I added the initialization code.

mhevery commented on packages/core/src/zone/ng_zone.ts:
```diff
@@ -324,9 +326,12 @@ function delayChangeDetectionForEvents(zone: NgZonePrivate) {
     if (!zone.fakeTopEventTask) {
       zone.fakeTopEventTask = Zone.root.scheduleEventTask('fakeTopEventTask', () => {
+        // Set isRunningDelayChangeDetection flag to avoid double entry
+        zone.isRunningDelayChangeDetection = true;
```
Rename to `isDelayChangeDetectionRunning`

JiaLiPassion replied:
Got it!

mhevery commented on packages/core/src/zone/ng_zone.ts:
```diff
@@ -290,8 +291,9 @@ interface NgZonePrivate extends NgZone {
-  if (zone._nesting == 0 && !zone.hasPendingMicrotasks && !zone.isStable) {
+function checkStable(zone: NgZonePrivate, ignoreRunningDelayChangeDetection = false) {
+  if (zone._nesting == 0 && !zone.hasPendingMicrotasks && !zone.isStable &&
+      (ignoreRunningDelayChangeDetection || !zone.isRunningDelayChangeDetection)) {
```
I am confused about why we need to have both `ignoreRunningDelayChangeDetection` and `isRunningDelayChangeDetection`. It is hard to follow.

I am wondering if moving `isRunningDelayChangeDetection` to here would better solve this issue. We can discuss in person if what I am asking does not make sense. 

It seems weird to me that `isRunningDelayChangeDetection` is set to `true` but then `chackStable` is called with `ignoreRunningDelayChangeDetection`

JiaLiPassion replied:
Yes, you are right, this is confusing, I removed the `ignoreRunningDelayChangeDetecition` parameter and only use the `isRunningDelayChangeDetection` flag here.

mhevery reviewed (changes requested):
Also can you add some jsdocs explaining the purpose of these flags?

JiaLiPassion commented:
@mhevery , I updated the PR and added the comment to explain the flag, please review, thanks!

mhevery commented on packages/core/src/zone/ng_zone.ts:
```diff
@@ -243,6 +244,17 @@ interface NgZonePrivate extends NgZone {
   hasPendingMicrotasks: boolean;
   lastRequestAnimationFrameId: number;
+  /**
+   * A flag to indicate whether currently is running delayed
+   * change detection, since when beginning to run the delayed
+   * change detection, it triggers onMicrotaskEmpty event and
+   * ApplicationRef call another ngZone.run() which tries to run
+   * another delayed change detection, so we need this flag to
+   * avoid re-entry.
+   * For detail, please refer here,
+   * https://github.com/angular/angular/pull/40540
```
Can we make the comment clearer:

---

A flag to indicate if NgZone is currently inside delayed change detection and to prevent re-entry. The flag is needed because it is possible to invoke the change detection from within change detection leading to incorrect behavior. 

---

Some things to keep in mind.
-- Keep sentences short. (shorter the better)
-- Try to avoid words as 'it' (antecedent) as it is unclear to the reader what exactly they refer to. Always expand them to the actual noun.

JiaLiPassion replied:
Got it, thanks, I will try to make comment more clearly.

mhevery commented on packages/core/src/zone/ng_zone.ts:
```diff
@@ -243,6 +244,15 @@ interface NgZonePrivate extends NgZone {
+  /**
+   * A flag to indicate if NgZone is currently inside checkStable and to prevent
+   * re-entry. The flag is needed because it is possible to invoke the change detection from within
+   * change detection leading to incorrect behavior.
```
This comment needs to be re-formatted so that all lines are equal in length.

JiaLiPassion replied:
Got it, I just updated it.

mhevery commented on packages/core/src/zone/ng_zone.ts:
```diff
@@ -304,12 +316,13 @@ function checkStable(zone: NgZonePrivate) {
 
 function delayChangeDetectionForEvents(zone: NgZonePrivate) {
-  if (zone.lastRequestAnimationFrameId !== -1) {
+  if (zone.isCheckStableRunning || zone.lastRequestAnimationFrameId !== -1) {
```
Why do we need `isCheckStableRunning` flag check here? Could we add a comment?

JiaLiPassion replied:
Got it, I have updated the comment here.
The reason we also need to check it here is when `shouldCoalesceRunChangeDetection` is true, we want the following case only run change detection once.

```
ngZone.run(() => {});
ngZone.run(() => {});
```

Since in this case, `checkStable()` is run in async way (requestAnimationFrame), so only check the flag in `checkStable()` will not prevent the `delayChangeDetectionForEvents` running twice. So we also need to check the flag here to prevent the async checkStable running multiple times too.

mhevery reviewed (approved):
Yes, I like this approach much better. Way cleaner and easier to follow.

mhevery commented:
[presubmit](http://test/OCL:354208379:BASE:354208401:1611796962641:9a60ebd)

mhevery commented:
Reverting due to `g3` breakage

JiaLiPassion committed:
fix(core): NgZone coaleascing options should trigger onStable correctly

fix https://github.com/angular/components/issues/21674

When setting `ngZoneRunCoalescing` to true, `onStable` is not emitted correctly.
the reason is before this commit, the code looks like this:

```
// application code call `ngZone.run()`
ngzone.run(() => {}); // step 1

// inside NgZone, in the OnInvoke hook, NgZone try to delay the checkStable()

function delayChangeDetectionForEvents(zone: NgZonePrivate) {
  if (zone.lastRequestAnimationFrameId !== -1) { // step 9
    return;
  }
  zone.lastRequestAnimationFrameId = zone.nativeRequestAnimationFrame.call(global, () => { // step 2
    if (!zone.fakeTopEventTask) {
      zone.fakeTopEventTask = Zone.root.scheduleEventTask('fakeTopEventTask', () => {
        zone.lastRequestAnimationFrameId = -1; // step 3
        updateMicroTaskStatus(zone); // step 4
        checkStable(zone); // step 6
      }, undefined, () => {}, () => {});
    }
    zone.fakeTopEventTask.invoke();
  });
  updatemicroTaskStatus(zone);
}

function updateMicroTaskStatus(zone: NgZonePrivate, ignoreCheckRAFId = false) {
  if (zone._hasPendingMicrotasks ||
      ((zone.shouldCoalesceEventChangeDetection || zone.shouldCoalesceRunChangeDetection) &&
       zone.lastRequestAnimationFrameId !== -1)) { // step 5
    zone.hasPendingMicrotasks = true;
  } else {
    zone.hasPendingMicrotasks = false;
  }
}

function checkStable(zone: NgZonePrivate) {
  if (zone._nesting == 0 && !zone.hasPendingMicrotasks && !zone.isStable) { // step 7
    try {
      zone._nesting++;
      zone.onMicrotaskEmpty.emit(null);
    ...
}

// application ref subscribe onMicroTaskEmpty
ngzone.onMicroTaskEmpty.subscribe(() => {
  ngzone.run(() => { // step 8
    tick();
  });
});

```

and the process is:
1. step 1: application call ngZone.run()
2. step 2: NgZone delay the checkStable() call in a requestAnimationFrame, and also set
zone.lastRequestAnimationFrameId
3. step 3: Inside the requestAnimationFrame callback, reset zone.lastRequestAnimationFrameId first
4. step 4: update microTask status
5, step 5: if zone.lastRequestAnimationFrameId is -1, that means no microTask pending.
6. step 6: checkStable and trigger onMicrotaskEmpty emitter.
7. step 7: ApplicationRef subscribed onMicrotaskEmpty, so it will call another `ngZone.run()` to process
tick()
8. step 8: And this new `ngZone.run()` will try to check `zone.lastRequestAnimationFrameId` in `step 9`
when trying to delay the checkStable(), and since the zone.lastRequestAnimationFrameId is already reset
to -1 in step 3, so this ngZone.run() will run into step 2 again.
9. and become a infinite loop..., so onStable is never emit

in this commit, there is a new flag `zone.isCheckStableRunning` added to
prevent re-entry when `shouldCoaleascing` flag is enabled.

JiaLiPassion commented:
@mhevery , I updated the PR to only resolve the issue when `ngZoneRunCoalescing` or `ngZoneEventCoalescing` enabled, and I will create another PR to resolve the `checkStable()` re-entry issue. Could you re-run the `presubmit`? Thanks!

mhevery commented:
[presubmit](http://test/OCL:354208379:BASE:357242777:1613160944469:2209d95e)

petebacondarwin commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -21,7 +21,7 @@
       "uncompressed": {
         "runtime-es2015": 3153,
-        "main-es2015": 432078,
+        "main-es2015": 432587,
```
I am surprised that this would add > 500 bytes. The new source code is basically only three lines which are each <100 chars. Perhaps the `master` branch was higher than 432078 before this PR?

JiaLiPassion replied:
@petebacondarwin , yes ,before this PR, the size is `432511`, so the real diff of this PR is `66` bytes.

mhevery commented:
[deflake1](https://test.corp.google.com/ui#id=OCL:354208379:BASE:357242779:1613160944514:dd3de33d)
[deflake2](https://test.corp.google.com/ui#id=OCL:354208379:BASE:357242779:1613511268994:579386e4)

JiaLiPassion commented:
@mhevery , I have rebased the PR, please review. Thanks!

atscott commented:
FYI - I reverted the change to the size tracking golden since there was only a newline change at the end of the file after a rebase. That removes the need for the size tracking approval.

atscott commented:
caretaker note: Zones change that was previously reverted. Merge & sync with caution
